src/rust/engine/workunit_store/src/lib.rs

```diff
@@ -853,6 +905,20 @@ impl RunningWorkunit {
     }
   }
 
+  ///
+  /// Marks the workunit as being blocked until the returned token is dropped.
+  ///
+  pub fn blocking(&mut self) -> BlockingWorkunitToken {
+    let mut token = BlockingWorkunitToken(None);
+    if let Some(ref mut workunit) = self.workunit {
+      if let WorkunitState::Started { blocked, .. } = &mut workunit.state {
+        blocked.store(true, atomic::Ordering::Relaxed);
```

Comment:
To ensure callers don't call `.blocking()` twice, maybe use `.compare_exchange` instead of `.store` and panic if the workunit was already marked as blocked?

Reply:
I also assume that multiple callers cannot or must not call `.blocking` concurrently?

Reply:
We're defended against concurrent access to this method because it takes the `RunningWorkunit` as `&mut self`.

But you're right that the same task could call `.blocking()` multiple times. Unless this method becomes much more widely used, I don't think it's worth panic'ing though.